added functionality to timer component

winnmar opened the pull request:
Known issue:
-Flag that controls disabling buttons does not work properly
-Needs to show more than miliseconds elapsed

winnmar committed:
added functionality to timer component

Known issue:
-Flag that controls disabling buttons does not work properly
-Needs to show more than miliseconds elapsed

bartlomiejzuber commented on src/components/Timer.js:
```diff
@@ -3,29 +3,63 @@ import React from 'react';
+        this.startClock = this.startClock.bind(this);
+        this.stopClock = this.stopClock.bind(this);
+
+        this.isStarted = false;
```
powinno byc w `state` inaczej react nie przeladuje komponentu gdy zmienisz ta wartosc, nie bedzie jej po prostu "widział"

bartlomiejzuber commented on src/components/Timer.js:
```diff
@@ -3,29 +3,63 @@ import React from 'react';
-        this.state = {hours: 1, minutes: 0, seconds: 0};
+        this.state = {beginTime: new Date(), endTime: new Date(), timeElapsed: 0};
 
         this.resetClock = this.resetClock.bind(this);
+        this.incrementClock = this.incrementClock.bind(this);
+        this.startClock = this.startClock.bind(this);
+        this.stopClock = this.stopClock.bind(this);
```
zamiast bindowania stosuje sie arrow function (wtedy masz automatyczne bindowanie thisa)

bartlomiejzuber commented on src/components/Timer.js:
```diff
@@ -3,29 +3,63 @@ import React from 'react';
                 <div>
-                    Time spent: {this.state.hours} : {this.state.minutes > 10 ? this.state.minutes : `0${this.state.minutes}`} : {this.state.seconds > 10 ? this.state.seconds : `0${this.state.seconds}`}
+                    Time spent: {this.state.timeElapsed}
                 </div>
-                <button onClick={this.resetClock}>
-                    Reset
-                </button>  
+                    <button disabled={this.isStarted} onClick={this.startClock}>
+                        Start
+                    </button> 
+                    <button disabled={!this.isStarted} onClick={this.stopClock}>
+                        Stop
+                    </button>
+                    <button onClick={this.resetClock}>
+                        Reset
+                    </button>
             </div> 
```
Ten komponent zawiera przyciski oraz wyswietla upływajacy czas co powoduje ze jezeli zaktualizujesz jego stan to reactjs wyrenderuje calosc gdy tak naprawde zmienia sie tylko czas. W takich sytuacjach wydziela się logike do oddzielnych komponentów. `TimeSpent` jak i `Buttons` powinny byc oddzielnymi komponentami.

bartlomiejzuber reviewed (commented):
I tak a propos klasowych kompo juz sie nie stosuje, wszystko nowe powinienes pisac uzywajac funkcyjnych kompo + hookow. Klasowe kompo były trudne w optymalizacji i dlatego zostały "porzucone", tzn. nadal są jednak nie polecaja ich uzywac w nowym kodzie .

winnmar committed:
changes after review

- Removed bind and added arrow functions instead
- Added smaller components instead of one big do-all
-

bartlomiejzuber commented on src/components/TimerDisplay.js:
```diff
@@ -0,0 +1,11 @@
+import React, { useState } from 'react';
+
+const TimerDisplay = (props) =>{   
+    return(
+        <div>
+            Time spent: {props.timeElapsed}
+        </div>
+    )
+}
```
```suggestion
const TimerDisplay = ({timeElapsed}) => (
        <div>
            Time spent: {timeElapsed}
        </div>
    )
```

Try to make code as much DRY as you can. :)

winnmar replied:
Możesz mi proszę powiedzieć co narusza tę zasadę? Rozumiem, że skoro ten komponent w swojej wersji zwraca tylko jednego diva ze zemitowaną wartością, to można to uprościć tak jak opisałeś, ale w którym miejscu tam się powtarzam?

bartlomiejzuber replied:
Chodzi o uproszczenie kodu, gdybyś chciał stworzyć nowy komponent na podstawie tego i w tym celu wykorzystał kopiuj/wklej, to byś miał już powtarzający się zbędny kod, stąd też komentarz o DRY.

bartlomiejzuber commented on src/components/TimerDisplay.js:
```diff
@@ -0,0 +1,11 @@
+    )
+}
+
+export default TimerDisplay;
```
https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad
https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/
https://www.reddit.com/r/webdev/comments/7j9nvy/why_we_have_banned_default_exports_in_javascript/

W skrócie, od wszystkiego są wyjąki, jednak w większości przypadków `default` export będzie stwarzał problemy tj. gorszy DX czy ręczna aktualizacja plików `index.ts`/`index.js`. Osobiście kieruję się zasadą korzystania wszędzie tam gdzie mogę z `named` exportów. Natomiast `default` exportów używam tam gdzie wymaga tego ficzer/konwencja, np. przy `lazy` loadingu.

winnmar replied:
Jak w takim razie eksportować na przykładzie `TimerDisplay`?

bartlomiejzuber replied:
Przy deklaracji komponentu dopisz `export` i to wystarczy, potem w `index.ts/js` używasz takiej konwencji: `export * from './TimerDisplay'`.

Żeby manualnie nie aktualizować indexów możesz użyć `https://www.npmjs.com/package/create-ts-index` i `https://github.com/paulmillr/chokidar`.

bartlomiejzuber commented on src/components/TimerControlButtons.js:
```diff
@@ -0,0 +1,25 @@
+
+const TimerControlButtons = ({props, onStartClick, onResetClick}) =>{
+
+    function handleStartClick(event){
+        onStartClick(event.target.name);
+    }
```
Używaj arrow function, od powstania es6 minęło już 5 lat, w tych czasach jest to standard.
```suggestion
    const handleStartClick = (event) => onStartClick(event.target.name);
```

Jakbyś chciał żeby ten komponent miał wydajność równą komponentowi klasowemu to powinieneś użyć również hooka `useCallback`:
```suggestion
    const handleStartClick = React.useCallback((event) => onStartClick(event.target.name), [onStartClick]);
```
https://reactjs.org/docs/hooks-reference.html#usecallback

bartlomiejzuber commented on src/components/Timer.js:
```diff
@@ -1,36 +1,33 @@
-    resetClock(){
+    resetClock = () => {
         this.setState(state => ({
             beginTime: new Date(), 
```
Przy starcie ustawiasz `beginTime` na `new Date()`, przy resecie robisz to ponownie. Wyglada na zbędną linie, mały szczegół ale rzuca się w oczy.

winnmar committed:
Update src/components/TimerControlButtons.js

Co-Authored-By: Bartłomiej Żuber <bajtas@gmail.com>
